### DISCUSSION TITLE
Error running a Gtk head on a Rpi device (XInitThreads)

### DISCUSSION THREAD

**Discussion by mcNets** [unoplatform/uno#12393] [Q&A] [answered]
Till now I was working in a Skia.Gtk project created the previous templates that properly runs in a Rpi device.

I was trying to use the new templates with Net 7 in the same device and I'm getting this error:

```
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
dotnet: ../../src/xcb_io.c:269: poll_for_event: L’asserció «!xcb_xlib_threads_sequence_lost» ha fallat.
```
I used a blank project (WinUI + Skia) and this command to build it:

    dotnet publish --ignore-failed-sources -r linux-arm -c Release -f net7.0 --self-contained false

The device is able to run a net7 console app.

**Comment by jeromelaban** [comment]
Thanks for the report. We're not getting this particular issue, and it may be related to environment installation. Which uno version and linux distribution are you using?

**Reply by mcNets** [reply]
**Raspbian version**

```
PRETTY_NAME="Raspbian GNU/Linux 11 (bullseye)"
NAME="Raspbian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=raspbian
ID_LIKE=debian
```

It's a 32bits OS (I'll check the same device with a 64bit version)

** UNO Extension Version **

    4.8.12.15

**Reply by jeromelaban** [reply]
Thanks. Does it happen as well with a blank flavored app as well? (using `-preset=blank`)

**Reply by mcNets** [reply]
I created a new app using:

    dotnet new unoapp -preset=blank -o TestRpi

I moved to the Test.Skia.Gtk folder and I executed:

    dotnet publish --ignore-failed-sources -r linux-arm -c Release -f net7.0 --self-contained false

Then I copied the content of the `bin\Release\net7.0\linux-arm\publish\` folder to the device and I tried to execute it by using: 

    dotnet TestRpi.Skia.Gtk.dll

And this is the error:

```
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
dotnet: ../../src/xcb_io.c:269: poll_for_event: L’asserció «!xcb_xlib_threads_sequence_lost» ha fallat.
```

I see the initial frame window and then it raises the error.

NOTE: The old application still runs in this device.

**Reply by jeromelaban** [reply]
Which version of Uno is your older app using? (The package reference, specifically)

**Reply by mcNets** [reply]
```
  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Logging" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
    <PackageReference Include="Uno.Extensions.Hosting" Version="2.2.0" />
    <PackageReference Include="Uno.WinUI.RemoteControl" Version="4.6.19" Condition="'$(Configuration)'=='Debug'" />
    <PackageReference Include="Uno.UI.Adapter.Microsoft.Extensions.Logging" Version="4.6.19" />
    <PackageReference Include="Uno.WinUI.Skia.Gtk" Version="4.6.19" />
  </ItemGroup>
```

**Reply by jeromelaban** [reply]
Thanks. Does your older app still work when upgrading to 4.8.33 ?

**Reply by mcNets** [reply]
It works.
This is the project file:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType Condition="'$(Configuration)'=='Release'">WinExe</OutputType>
    <OutputType Condition="'$(Configuration)'=='Debug'">Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
    <ApplicationManifest>app.manifest</ApplicationManifest>
  </PropertyGroup>
  <ItemGroup Condition="exists('..\VP_Sysmac.Uno.Windows')">
    <EmbeddedResource Include="..\VP_Sysmac.Uno.Windows\Package.appxmanifest" LogicalName="Package.appxmanifest" />
    <Content Include="..\VP_Sysmac.Uno.Windows\Images\StoreLogo.png" Link="Assets\StoreLogo.png" />
    <Content Include="Assets\Fonts\uno-fluentui-assets.ttf" />
  </ItemGroup>
  <ItemGroup>
    <UpToDateCheckInput Include="..\VP_Sysmac.Uno.Shared\**\*.xaml" />
  </ItemGroup>
  <ItemGroup>
    <Content Include="..\VP_Sysmac.Uno.Shared\Assets\Fonts\Okuda-A5PL.ttf" Link="Assets\Fonts\Okuda-A5PL.ttf" />
  </ItemGroup>
  <ItemGroup>
    <PackageReference Include="Microsoft.Extensions.Logging" Version="7.0.0" />
    <PackageReference Include="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
    <PackageReference Include="Uno.Extensions.Hosting" Version="2.2.0" />
    <PackageReference Include="Uno.WinUI.RemoteControl" Version="4.6.19" Condition="'$(Configuration)'=='Debug'" />
    <PackageReference Include="Uno.UI.Adapter.Microsoft.Extensions.Logging" Version="4.6.19" />
    <PackageReference Include="Uno.WinUI.Skia.Gtk" Version="4.8.33" />
  </ItemGroup>
  <ItemGroup>
    <ProjectReference Include="..\..\VP_Sysmac.APIClient\VP_Sysmac.APIClient.csproj" />
    <ProjectReference Include="..\..\VP_Sysmac.Domain\VP_Sysmac.Domain.csproj" />
    <ProjectReference Include="..\..\VP_Sysmac.Gpio\VP_Sysmac.Gpio.csproj" />
    <ProjectReference Include="..\..\VP_Sysmac.Services\VP_Sysmac.Stores.csproj" />
  </ItemGroup>
  <Import Project="..\VP_Sysmac.Uno.Shared\VP_Sysmac.Uno.Shared.projitems" Label="Shared" />
</Project>
```

**Reply by mcNets** [reply]
I've tried a new blank project with Net6 and it runs.

By blank I mean no mobile projects.

Then I've updated the project to net7 and it fails again.

And then I've build it in debug and that's the output of the console:

```
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint [fe80::9fe:767a:d29e:33e6].
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint 192.168.1.137.
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint [fe80::af1e:7809:15d7:1975].
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint 169.254.136.28.
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint [fe80::e90e:352b:1043:3046].
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint 169.254.161.54.
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint [fe80::bedc:126:5169:1f35].
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint 192.168.1.144.
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint [fe80::6aba:f10e:923a:a42d].
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get remote control server port from the IDE for endpoint 169.254.61.166.
fail: Uno.UI.RemoteControl.RemoteControlClient[0]
      Failed to get any remote control server endpoint from the IDE.
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
dotnet: ../../src/xcb_io.c:269: poll_for_event: L’asserció «!xcb_xlib_threads_sequence_lost» ha fallat.
Avortat
```

This `192.168.1.137` the current IP of my laptop not the device.

**Reply by jeromelaban** [reply]
Thanks.

Could you try adding this to your gtk entry point (`Program.cs`):

```csharp
struct Native
{
	[DllImport("libX11.so")]
	internal static extern int XInitThreads();
}
```

Then invoke it like this:
```csharp
if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
{
   Native.XInitThreads();
}
```

If it really is a threading issue, there's definitely something we need to adjust.

**Reply by mcNets** [reply]
I've tried before and after `var host = new GtkHost(() => new AppHead(), args);`

```
namespace TestRpi6.Skia.Gtk {
        
    struct Native {
        [DllImport("libX11.so")]
        internal static extern int XInitThreads();
    }

    public class Program {
        public static void Main(string[] args) {
            ExceptionManager.UnhandledException += delegate (UnhandledExceptionArgs expArgs) {
                Console.WriteLine("GLIB UNHANDLED EXCEPTION" + expArgs.ExceptionObject.ToString());
                expArgs.ExitApplication = true;
            };

            var host = new GtkHost(() => new AppHead(), args);

            if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux)) {
                Console.WriteLine("XInitThreads");
                Native.XInitThreads();
            }

            host.Run();
        }
    }
}
```

That's the result:

```
XInitThreads
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
dotnet: ../../src/xcb_io.c:269: poll_for_event: L’asserció «!xcb_xlib_threads_sequence_lost» ha fallat.
Avortat
```
XInitThreads returns 1

**Reply by jeromelaban** [reply]
Thanks. Can you try placing it at the very beginning, just to be sure? Otherwise, we'll need to put the logging to tracing level to determine where this comes from.

**Reply by mcNets** [reply]
I didn't mention it, but I also tried from the beginning. 
Let me know if I can do anything else.

**Reply by mcNets** [reply]
I'm not sure if it is related but today I've used a device with a 64bits OS. And now I'm facing this issue:

```
Unhandled exception. System.IO.FileNotFoundException: Could not load file or assembly 'GLibSharp, Version=3.24.24.38, Culture=neutral, PublicKeyToken=null'. The system cannot find the file specified.

File name: 'GLibSharp, Version=3.24.24.38, Culture=neutral, PublicKeyToken=null'
Abortado
```

GLib Sharp.dll is located in the application folder and its version is 3.24.24.38

I published it using this command:

`dotnet publish --ignore-failed-sources -r linux-arm64 -c Release -f net7.0 --self-contained false`

**Reply by jeromelaban** [reply]
Thanks. Does it happen when you try running the output of the bin folder, using a `dotnet build` command?

**Reply by mcNets** [reply]
No, I build the app directly in the device's folder, and then I execute it by using `dotnet app_name.dll`

**Reply by jeromelaban** [reply]
Thanks, so this would mean that it's the publish command that fails to create a proper app layout.

**Reply by mcNets** [reply]
That's my powershell command to build 32 bits:

```
$ip = $args[0] 
dotnet publish --ignore-failed-sources -r linux-arm -c Release -f net7.0 --self-contained false -o \\$ip\PiShared\Sysmac /p:PublishSingleFile=false
```

And this 64 bits apps:

```
$ip = $args[0]  
dotnet publish --ignore-failed-sources -r linux-arm64 -c Release -f net7.0 --self-contained false -o \\$ip\PiShared\Sysmac /p:PublishSingleFile=false
```

I only change the net. release depending on the app. ( -f )

**Reply by jeromelaban** [reply]
I'm sorry, I'm missing some details in that last message. Are both not working, or just one?

**Reply by mcNets** [reply]
>Thanks, so this would mean that it's the publish command that fails to create a proper app layout.

Just to show you how I publish the app in the device. So, I publish from my Windows machine to the Linux folder.

**Reply by mcNets** [reply]
@jeromelaban I'm rewriting the old application that is already in production. I did a lot of work in the new one, obviously without checking if  it could run in the device, I took it for granted. Do you think there is any workaround to get it working? Neither net6 nor net7 is working. I could try to build the Skia.Gtk library but and I don't know where XInitThreads should be executed.

**Reply by jeromelaban** [reply]
I don't believe `XInitThreads` is the way to fix this issue. I'm not able to reproduce your issue either, whether it's through self contained or not. I'm not able to test with arm32 at this time, though. Could you try with arm64?

I tested with:
```
dotnet new unoapp -preset=blank -o TestRpi
dotnet publish --ignore-failed-sources -r linux-arm -c Release -f net7.0 --self-contained false
```

And the app is running fine.

**Reply by mcNets** [reply, marked as answer]
You are right, in a 64bits OS runs fine. (Hardware is the same)

Well I'm sorry Jérôme, thank you very much for the time you spent in this topic.

![image](https://github.com/unoplatform/uno/assets/24267381/6f9620e7-294b-46d3-853d-768f9540403b)

**Reply by jeromelaban** [reply]
No worries and thanks for the update, this still means that there's something strange happening when running on arm32.
